tests/test_utils.py

```diff
@@ -0,0 +1,58 @@
+import unittest
+from datetime import datetime
+import numpy as np
+from utils import (
+    calculate_ZCB_values, 
+    get_ZCB_vector
+)
+
+class TestCalculateZCBValue(unittest.TestCase):
+    """
+    Unit tests for the calculate_ZCB_value function.
+    """
+
+    def test_calculate_ZCB_value(self):
+        """
+        Test the calculation of zero-coupon bond (ZCB) value.
+        Verifies that the ZCB value is correctly computed for a given
+        discount rate, current date, and maturity date.
+        """
+        rates = [0.05, 0.04]
+        start_dates = [datetime(2025, 1, 1), datetime(2025, 1, 1)]
+        end_dates = [datetime(2026, 1, 1), datetime(2027, 1, 1)]
+        expected_values = [
+            np.exp(-rates[0] * 1.0),  # 1 year
+            np.exp(-rates[1] * 2.0)   # 2 years
```

Comment:
it think there is a misunderstanding here.  the expected value for year 2 should be:

exp( -(rates[0] * 1.0 + rates[1]* 1.0) )

every zcb should be a function of all rates that came before it.  I think what you have are two different integrals that overlap for the first year instead of a single integral over all the rates

Reply:
ZCB value is a lot different than ZCB vector. it was just a helper function for the old version of GetZCBVector. It is doing what it's supposed to. You can see in the test_calculate_ZCB_vector that the test is using your expression here

Reply:
are the years on lines 24 and 25 deltas so the total of time 3 years?  or offsets from time zero?